src/allocation_engine/interval_tree.rs

```diff
@@ -185,8 +185,173 @@ struct InnerNode {
+        self.left = new_root.right.take();
+        self.update_cached_height();
+        new_root.right = Some(self);
+        new_root.update_cached_height();
+        new_root
+    }
+
+    /// Performs a rotation when the left successor is too high.
+    #[allow(dead_code)]
+    fn rotate_left_successor(mut self: Box<Self>) -> Box<Self> {
+        let left = self.left.take().expect("Node is broken");
+        if height(&left.left) < height(&left.right) {
+            let rotated = left.rotate_left();
+            self.left = Some(rotated);
+            self.update_cached_height();
+        } else {
+            self.left = Some(left);
+        }
+        self.rotate_right()
+    }
+
+    /// Performs a rotation when the right successor is too high.
+    #[allow(dead_code)]
+    fn rotate_right_successor(mut self: Box<Self>) -> Box<Self> {
+        let right = self.right.take().expect("Node is broken");
+        if height(&right.left) > height(&right.right) {
+            let rotated = right.rotate_right();
+            self.right = Some(rotated);
+            self.update_cached_height();
+        } else {
+            self.right = Some(right);
+        }
+        self.rotate_left()
+    }
+
+    /// Deletes the entry point of this tree structure.
+    #[allow(dead_code)]
+    fn delete_root(mut self) -> Option<Box<Self>> {
+        match (self.left.take(), self.right.take()) {
+            (None, None) => None,
+            (Some(l), None) => Some(l),
+            (None, Some(r)) => Some(r),
+            (Some(l), Some(r)) => Some(Self::combine_subtrees(l, r)),
+        }
+    }
+
+    /// Finds the minimal key below the tree and returns a new optional tree
+    /// where the minimal value has been removed and the (optional) minimal node
+    /// as tuple (min_node, remaining).
+    #[allow(dead_code)]
+    fn get_new_root(mut self: Box<Self>) -> (Box<Self>, Option<Box<Self>>) {
+        match self.left.take() {
+            None => {
+                let remaining = self.right.take();
+                (self, remaining)
+            }
+            Some(left) => {
+                let (min_node, left) = left.get_new_root();
+                self.left = left;
+                (min_node, Some(self.updated_node()))
+            }
+        }
+    }
+
+    /// Creates a single tree from the subtrees resulted from deleting the root
+    /// node.
+    #[allow(dead_code)]
+    fn combine_subtrees(l: Box<Self>, r: Box<Self>) -> Box<Self> {
+        let (mut new_root, remaining) = r.get_new_root();
+        new_root.left = Some(l);
+        new_root.right = remaining;
+        new_root.updated_node()
+    }
+
+    /// Updates cached information of the node.
+    #[allow(dead_code)]
+    fn update_cached_height(&mut self) {
+        self.height = max(height(&self.left), height(&self.right)) + 1;
+    }
+
+    /// Updates the sub-tree to keep balance.
+    #[allow(dead_code)]
+    fn updated_node(mut self: Box<Self>) -> Box<Self> {
```

Comment:
not sure the meaning of `updated` here, maybe this should be `update_node`

Reply:
s/updated_node/update_node